### PR TITLE
API overhaul

### PR DESCRIPTION
**Pull request by jdef**
* Bump to go1.7 and begin integrating support for `context.Context`
* Simplify "core" APIs and move the rest to `extras`.
* Introduce "rules" in favor of "Decorators"
* Code-generation instead of copy/paste code (e.g. scheduler/executor call/event handling APIs)
* Begin to internalize error types that don't have great recovery semantics
* Extract bits from `api/v1/lib/{operations,resources}` modules into separate packages:
  * `resourcefilters`, `resources`, `resourcetest`, and `scheduler/operations`
  * lots of refactoring here
* `apierrors` cleanup, introduce `Temporary` and `SubscriptionLoss` methods
* `encoding` and `recordio` cleanup and unit testing
  - refactor `Codec` API and introduce `MediaType`
* new `debug` package (TODO: maybe move under /internal?)
* `operations` API cleanup

TODO:
- [x] tag master with v0.0.3 before merging this
- [mostly] unit tests for new code (including generated code)
  - [x] `RateLimit` needs test for different `Overflow` types
  - ~[ ] `EveryN` needs tests~
- [x] clean up code generation (e.g. prefer `-type` to `-event_type`)
- [x] go-generate rules for executor events?
- [x] go-generate handler/rule integration?

### CODE REVIEW

File: api/v1/lib/offers.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package mesos
+
+// Offers is a convenience type wrapper for a collection of mesos Offer messages
+type Offers []Offer
```
Comment (jdef) [inline api/v1/lib/offers.go]: this is questionable

File: api/v1/lib/builders.go [hunk 2]
```diff
@@ -2,49 +2,65 @@ package mesos
 
 type (
```
Comment (jdef) [inline api/v1/lib/builders.go]: not sure that this file belongs in the mesos package, maybe in a `builders` package (stuffed into `extras`)?

File: api/v1/lib/extras/offers/offers.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
+package offers
```
Comment (jdef) [inline api/v1/lib/extras/offers/offers.go]: (a) this is begging for type-generation
(b) offers are a scheduler concern, consider moving to `scheduler/offers`

File: api/v1/lib/extras/scheduler/controller/controller.go [hunk 4]
```diff
@@ -9,79 +9,111 @@ import (
...
+	Option func(*Config) Option
 
-		// FrameworkID returns the current Mesos-assigned framework ID. Frameworks are expected to
-		// track this ID (that comes from Mesos, in a SUBSCRIBED event).
-		FrameworkID() string
-
-		// Error is an error handler that is invoked at the end of every subscription cycle; the given
-		// error may be nil (if no errors occurred).
-		Error(error)
+	// Config is a controller configuration. Public fields are REQUIRED. Optional properties are
```
Comment (jdef) [inline api/v1/lib/extras/scheduler/controller/controller.go]: there are no more public fields, fix this comment

File: api/v1/lib/extras/store/singleton.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+	SingletonDecorator interface {
+		Decorate(Singleton) Singleton
+	}
+)
+
+func (f GetFunc) Get() (string, error) { return f() }
+func (f SetFunc) Set(s string) error   { return f(s) }
+
+var ErrNotFound = errors.New("value not found in store")
```
Comment (jdef) [inline api/v1/lib/extras/store/singleton.go]: not crazy about this. maybe a "not found" condition should just return empty string?

File: api/v1/lib/scheduler/events/events.go [hunk 6]
```diff
@@ -133,53 +159,71 @@ func DefaultHandler(eh Handler) Option {
...
+	Cause error
+}
+
+func (err *AckError) Error() string { return err.Cause.Error() }
+
 // AcknowledgeUpdates generates a Handler that sends an Acknowledge call to Mesos for every
-// UPDATE event that's received.
-func AcknowledgeUpdates(callerGetter func() calls.Caller) Handler {
+// UPDATE event that's received (that requests an ACK).
+func AcknowledgeUpdates(callerLookup func() calls.Caller) Handler {
```
Comment (jdef) [inline api/v1/lib/scheduler/events/events.go]: redundant w/ the rule version. can we refactor this leverage the Rule implementation?

File: api/v1/cmd/msh/msh.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+func buildClient() calls.Caller {
+	return httpsched.NewCaller(httpcli.New(
+		httpcli.Endpoint(fmt.Sprintf("http://%s/api/v1/scheduler", MesosMaster)),
+	))
+}
+
+func buildEventHandler(caller calls.Caller) events.Handler {
+	logger := controller.LogEvents()
+	return controller.LiftErrors().Handle(events.Handlers{
+		scheduler.Event_FAILURE:    logger,
```
Comment (jdef) [inline api/v1/cmd/msh/msh.go]: redundant line, logger is already invoked by Otherwise

File: api/v1/cmd/msh/msh.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	}
+	if err := run(); err != nil {
+		if exitErr, ok := err.(ExitError); ok {
+			if code := int(exitErr); code != 0 {
+				log.Println(exitErr)
+				os.Exit(code)
+			}
+			// else, code=0 indicates success, exit normally
+		} else {
+			panic(fmt.Sprintf("%#v", err))
```
Comment (jdef) [inline api/v1/cmd/msh/msh.go]: log.Fatal instead?

File: api/v1/lib/encoding/codec.go [hunk 9]
```diff
@@ -10,45 +11,78 @@ import (
...
 )
 
-var (
-	// ProtobufCodec is the Mesos scheduler API Protobufs codec.
-	ProtobufCodec = Codec{
-		Name:       "protobuf",
-		MediaTypes: [2]string{ProtobufMediaType, ProtobufMediaType},
-		NewEncoder: NewProtobufEncoder,
-		NewDecoder: NewProtobufDecoder,
+var DefaultCodecs = func() map[MediaType]Codec {
```
Comment (jdef) [inline api/v1/lib/encoding/codec.go]: get rid of the func

File: api/v1/lib/extras/resources/reduce.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
+package resources
+
+import (
+	"github.com/mesos/mesos-go/api/v1/lib"
+	"github.com/mesos/mesos-go/api/v1/lib/resourcefilters"
+)
+
+// Reducer applies some operation to possibly combine an accumulator (acc) and a resource (x), returning
```
Comment (jdef) [inline api/v1/lib/extras/resources/reduce.go]: s/possibly combine/evaluate/

Comment (jdef) [inline api/v1/lib/extras/resources/reduce.go]: think/talk about mutability of acc

File: api/v1/lib/extras/scheduler/operations/operations.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+		}
+		stripped := proto.Clone(&volumes[i]).(*mesos.Resource)
+		stripped.Disk = nil
+		result.Subtract1(volumes[i])
+		result.Add1(*stripped)
+	}
+	return result, nil
+}
+
+func Apply(operation *mesos.Offer_Operation, resources mesos.Resources) (mesos.Resources, error) {
```
Comment (jdef) [inline api/v1/lib/extras/scheduler/operations/operations.go]: do we really need the `mesos.Resources` type here?

File: api/v1/lib/recordio/reader.go [hunk 12]
```diff
@@ -4,94 +4,168 @@ import (
...
 	"github.com/mesos/mesos-go/api/v1/lib/encoding/framing"
 )
 
-// NewReader returns an io.Reader that unpacks the data read from r out of
-// RecordIO framing before returning it.
-func NewReader(r io.Reader) io.Reader {
-	br, ok := r.(*bufio.Reader)
-	if !ok {
-		br = bufio.NewReader(r)
+type Debugger bool
```
Comment (jdef) [inline api/v1/lib/recordio/reader.go]: extract this into a debug package

File: api/v1/lib/resourcefilters/resources.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
+package resourcefilters
```
Comment (jdef) [inline api/v1/lib/resourcefilters/resources.go]: rename this file to `resourcefilters.go`

File: api/v1/lib/resources.go [hunk 14]
```diff
@@ -89,291 +54,36 @@ func (t ResourceErrorType) Generate(reason string) error {
...
 		}
 	}
-	return &ResourceError{errorType: t, reason: msg}
+	return &resourceError{errorType: t, reason: msg}
 }
 
-func (err *ResourceError) Type() ResourceErrorType { return err.errorType }
-func (err *ResourceError) Reason() string          { return err.reason }
-func (err *ResourceError) Resource() Resource      { return err.spec }
+func (err *resourceError) Type() resourceErrorType { return err.errorType }
```
Comment (jdef) [inline api/v1/lib/resources.go]: this is silly, either expose `resourceErrorType` as a public type (and also the consts) or else nix this func
